sync_diff_inspector/chunk/chunk.go

```diff
@@ -0,0 +1,460 @@
+	if c.IsLast {
+		return true
+	}
+	// calculate from bounds
+	for _, b := range c.Bounds {
+		if b.HasUpper {
+			return false
+		}
+	}
+	return true
+}
+
+func (c *Range) IsFirstChunkForTable() bool {
+	if c.IsFirst {
+		return true
+	}
+	// calculate from bounds
+	for _, b := range c.Bounds {
+		if b.HasLower {
+			return false
+		}
+	}
+	return true
+}
+
+// String returns the string of Range, used for log.
+func (c *Range) String() string {
+	chunkBytes, err := json.Marshal(c)
+	if err != nil {
+		log.Warn("fail to encode chunk into string", zap.Error(err))
+		return ""
+	}
+
+	return string(chunkBytes)
+}
+
+func (c *Range) ToString(collation string) (string, []interface{}) {
+	if collation != "" {
+		collation = fmt.Sprintf(" COLLATE '%s'", collation)
+	}
+
+	/* for example:
+	there is a bucket in TiDB, and the lowerbound and upperbound are (A, B1, C1), (A, B2, C2), and the columns are `a`, `b` and `c`,
+	this bucket's data range is (a = A) AND (b > B1 or (b == B1 and c > C1)) AND (b < B2 or (b == B2 and c <= C2))
+	*/
+
+	sameCondition := make([]string, 0, 1)
+	lowerCondition := make([]string, 0, 1)
+	upperCondition := make([]string, 0, 1)
+	sameArgs := make([]interface{}, 0, 1)
+	lowerArgs := make([]interface{}, 0, 1)
+	upperArgs := make([]interface{}, 0, 1)
+
+	preConditionForLower := make([]string, 0, 1)
+	preConditionForUpper := make([]string, 0, 1)
+	preConditionArgsForLower := make([]interface{}, 0, 1)
+	preConditionArgsForUpper := make([]interface{}, 0, 1)
+
+	i := 0
+	for ; i < len(c.Bounds); i++ {
+		bound := c.Bounds[i]
+		if !(bound.HasLower && bound.HasUpper) {
+			break
+		}
+
+		if bound.Lower != bound.Upper {
+			break
+		}
+
+		sameCondition = append(sameCondition, fmt.Sprintf("%s%s = ?", dbutil.ColumnName(bound.Column), collation))
+		sameArgs = append(sameArgs, bound.Lower)
+	}
+
+	for ; i < len(c.Bounds); i++ {
+		bound := c.Bounds[i]
+		lowerSymbol := gt
+		upperSymbol := lt
+		if i == len(c.Bounds)-1 {
+			upperSymbol = lte
+		}
+
+		if bound.HasLower {
+			if len(preConditionForLower) > 0 {
+				lowerCondition = append(lowerCondition, fmt.Sprintf("(%s AND %s%s %s ?)", strings.Join(preConditionForLower, " AND "), dbutil.ColumnName(bound.Column), collation, lowerSymbol))
+				lowerArgs = append(append(lowerArgs, preConditionArgsForLower...), bound.Lower)
+			} else {
+				lowerCondition = append(lowerCondition, fmt.Sprintf("(%s%s %s ?)", dbutil.ColumnName(bound.Column), collation, lowerSymbol))
+				lowerArgs = append(lowerArgs, bound.Lower)
+			}
+			preConditionForLower = append(preConditionForLower, fmt.Sprintf("%s%s = ?", dbutil.ColumnName(bound.Column), collation))
+			preConditionArgsForLower = append(preConditionArgsForLower, bound.Lower)
+		}
+
+		if bound.HasUpper {
+			if len(preConditionForUpper) > 0 {
+				upperCondition = append(upperCondition, fmt.Sprintf("(%s AND %s%s %s ?)", strings.Join(preConditionForUpper, " AND "), dbutil.ColumnName(bound.Column), collation, upperSymbol))
```

Comment:
below we use string template `(%s) AND (%s)`, but here `(%s AND %s)`

Reply:
for example, ((a > 0) OR `(a = 0 AND b > 1)`) AND ((a < 1) OR (a = 1 AND b <= 2))
`(a = 0 AND b > 1)` has not any other operators such as `OR`. So it can use template `(%s AND %s)`  directly.
the `%s` in `(%s) AND (%s)` may have `OR` operators when the chunk has multi bounds.